Fix webpack compatibility issue

grimmer0125 opened the pull request:
This is to solve 
1) #1875 for Creat React App environment and **2.** environment
2) Fix webpack building error when using `npx webpack` with its default and minimal setting for your front-end code. Error: import("path") and import("vm") & **1.**
3) after fixing **1 & 2**, `import` of loadScript` in the browser will throw `module not found` error since the `import` function is already polluted by `webpack`. It is a runtime issue. 

Besides this PR, `Pydodie npm` also needs #1849 to make it work.

grimmer0125 commented:
> Could you clone your repo and build it with webpack in as [a new run step in test-js](https://github.com/pyodide/pyodide/blob/main/.circleci/config.yml#L226) ?

I have tried to add the failure case (before the fix) and switched to the successful case (after the fix). Currently, I just copy the fixed `load-script.js` file from this pyodide repo into the cloned repo. It can change to use symbolic link the whole folder or new published pyodide package. 

I think the following steps are
1. organize the commits of my repo 
2. move it in Pyodide organization 
3. Change the used GitHub URL in the `config.yml` in this PR 
4. Merge the PR. 
5. Release a new npm 
6. (optional) change to use new pyodide npm package instead 

@rth does this make sense? 

The test on the fixed runtime issue (`loadScript = async (url) => await import(/* webpackIgnore: true */ url);` could be another PR if need.

grimmer0125 commented:
> @grimmer0125 I think we prefer if you use merge rather than rebase because the force pushes confuse the github review tools.

Got it. I forget I can use other branches to do the circleci experienemcne.

hoodmane commented:
> I forget I can use other branches to do the circleci experiments.

Well if there is a good reason force pushes are okay. Just try to avoid if practical.

rth commented on .circleci/config.yml:
```diff
@@ -224,6 +224,18 @@ jobs:
+            cp ../src/js/load-pyodide.js node_modules/pyodide/load-pyodide.js                  
+            head -20 node_modules/pyodide/load-pyodide.js
+            npx webpack
+            cd ..
+            rm -rf webpack-pyodide-demo          
```
```suggestion         
```
No need for this, the docker instance will be destroyed anyway.

rth commented:
Thanks a lot for doing this @grimmer0125 !

> organize the commits of my repo

I wouldn't worry too much about that, people don't look at commits history that much, but as you prefer. What we would also need however is to add a README explaining what that this repo is and if it's only for testing or if users can use it as an example. Also please add a LICENSE. The readme should also explain why there is a `webpack-pyodide-demo/node_modules/pyodide/load-pyodide.js` file and I assume that it would be removed after the next Pyodide release? Maybe let's add those via a PR so we can discuss it there?

> move it in Pyodide organization

Could you please add me as as admin to that repo? I'll move it? You will still keep your permissions on it.

grimmer0125 commented:
> I wouldn't worry too much about that, people don't look at commits history that much, but as you prefer. What we would also need however is to add a README explaining what that this repo is and if it's only for testing or if users can use it as an example. Also please add a LICENSE. The readme should also explain why there is a webpack-pyodide-demo/node_modules/pyodide/load-pyodide.js file and I assume that it would be removed after the next Pyodide release? Maybe let's add those via a PR so we can discuss it there?

I've added README, LICENSE file, and note it is a minimal example project. To avoid things complicated, I removed the committed `load-pyodide.js` and it should be installed from npm package. So now this project does not work byt itself. And the CircleCI will copy `load-pyodide.js` to the downloaded [pyodide-webpack-example] (https://github.com/grimmer0125/pyodide-webpack-example) to make the test sucessfully. If you think it is better to keep `pyodide-webpack-example` running successfully by default now and we can add the fixed `load-pyodide.js` back.

>  Could you please add me as as admin to that repo? I'll move it? You will still keep your permissions on it.
I've added you as a collaborator of my [pyodide-webpack-example project](https://github.com/grimmer0125/pyodide-webpack-example).

rth commented:
> I've added you as a collaborator of my pyodide-webpack-example project.

I received the invite, but that only gives write permission, not admin ones so I'm not able to move it. Other maybe could you transfer it to me and I'll transfer it there? 

>  After a new Pyodide npm package is released, pyodide-webpack-example and Pyodide CircleCI can just use npm version of Pyodide to run well. 

Yes, let's do that. We can very well release a 0.18.3 to npm just with those fixes, without waiting for a complete Pyodide release. Though then the question how to keep a changelog in those conditions.

grimmer0125 commented:
> I received the invite, but that only gives write permission, not admin ones so I'm not able to move it. Other maybe could you transfer it to me and I'll transfer it there?

I did not find out any admin role setting in a personal repo and it seems to only exist in a GitHub organization. I've started to transfer it to you and please check.

rth commented:
@grimmer0125 the repository was moved.

rth commented:
Please also add a changelog entry (preferable in a new commit without force-pushing :) )  . Otherwise LGTM. Thanks!

rth commented on .circleci/config.yml:
```diff
@@ -224,6 +224,17 @@ jobs:
+            cp ../src/js/load-pyodide.js node_modules/pyodide/load-pyodide.js                  
+            head -20 node_modules/pyodide/load-pyodide.js
+            npx webpack
+            cd ..
+            rm -rf pyodide-webpack-example          
```
```suggestion
   
```

grimmer0125 commented:
I have added a changlog.

hoodmane commented:
Thanks @grimmer0125!
